src/Microsoft.XmlSerializer.Generator/src/Sgen.cs

```diff
@@ -467,5 +475,77 @@ private static string GetTempAssemblyName(AssemblyName parent, string ns)
         {
             return parent.Name + ".XmlSerializers" + (ns == null || ns.Length == 0 ? "" : "." + ns.GetHashCode());
         }
+
+        private static void ParseReferences(string value, Dictionary<string, string> dictionary)
```

Comment:
This can essentially be re-written to avoid a lot of the allocations and be faster, with the use of Span.

As an aside, you don't need a List here if you do everything within a single loop.

```C#
    private static void ParseReferencesNew(string value, Dictionary<string, string> dictionary)
    {
        if (value.Length > 0)
        {
            string[] entries = value.Split(';');
            for (int i = 0; i < entries.Length; i++)
            {
                ReadOnlySpan<char> reference = entries[i].AsSpan().Trim();
                if (reference.Length == 0)
                    continue;

                if (reference.EndsWith(".dll") || reference.EndsWith(".exe"))
                {
                    if (File.Exists(reference)) // need to add File.Exists overload that takes ReadOnlySpan<char>
                    {
                        ReadOnlySpan<char> filename = Path.GetFileNameWithoutExtension(reference);
                        if (filename.Length != 0)
                        {
                            dictionary.Add(new string(filename), new string(reference));
                        }
                    }
                }
            }
        }
    }
```

Can we add a File.Exists overload that takes in a `ReadOnlySpan<char>`?
https://github.com/dotnet/corefx/blob/a00ae9072b9244d2fa7ee282ec90a5d8500704ac/src/System.IO.FileSystem/src/System/IO/File.cs#L172
cc @JeremyKuhne, @jkotas

Reply:
There is already some work happening in that area: https://github.com/dotnet/corefx/pull/27347

Reply:
Thanks for suggestion. I just have a try,  but seems the ReadOnlySpan is still not declared yet.  We can consider your way later.

Reply:
> Thanks for suggestion. I just have a try, but seems the ReadOnlySpan is still not declared yet. We can consider your way later.

I am not sure why you don't have access to ReadOnlySpan, but that is totally fine. It is isn't necessary here and can be re-considered later if relevant.

> As an aside, you don't need a List here if you do everything within a single loop.

That said, I think we can still avoid a lot of redundant work here by doing everything within a single loop (or better yet, do both ParseReferences and SgenAssemblyResolver in a single method which avoids the need to generate a dictionary). See my other comment -
 https://github.com/dotnet/corefx/pull/27425#discussion_r170434450

Reply:
Something like this (note I removed the use of Span, just using strings here, and now there is no need for a ParseReferences method):

```C#
private static Assembly SgenAssemblyResolver(object source, ResolveEventArgs e)
{
    try
    {
        if (string.IsNullOrEmpty(s_references) || string.IsNullOrEmpty(e.Name))
            return null;
        
        string[] eventArgs = e.Name.Split(',');
        
        if (eventArgs.Length == 0)
            return null;
        
        string assemblyname = eventArgs[0];
        if (string.IsNullOrEmpty(assemblyname))
            return null;
        
        string[] entries = s_references.Split(';');
        for (int i = 0; i < entries.Length; i++)
        {
            string reference = entries[i].Trim();
            if (reference.Length == 0)
                continue;

            if ((reference.EndsWith(".dll") || reference.EndsWith(".exe"))
                && File.Exists(reference))
            {
                string filename = Path.GetFileNameWithoutExtension(reference);
                if (filename.Length != 0 && filename == assemblyname)
                    return Assembly.LoadFrom(reference);
            }
        }
    }
    catch (Exception exp)
    {
        if (exp is ThreadAbortException || exp is StackOverflowException || exp is OutOfMemoryException)
        {
            throw;
        }

        WriteWarning(exp, true);
    }

    return null;
}
```